### PR TITLE
Pipewire: constify methods 

### PR DESCRIPTION
**Pull request by lrusak**
This adds const to a bunch of pipewire methods. I'm not 100% sure on the mutable members or if that is worth doing.

I also added two commits to fix the use of explicit.

### CODE REVIEW

**Comment by lrusak** [comment]
jenkins build this please

**Comment by neo1973** [comment]
>I'm not 100% sure on the mutable members

I think you shouldn't do it, returning a mutable reference from a const method is very sketchy anyway. If there are cases where const access would be beneficial make const overloads that also return const references.

I also think that most of the methods in this PR should not be const even if the type system doesn't prevent it. E.g. `CPipewireStream::Connect` or `CPipewireStream::SetActive` clearly change the state of the object and therefore shouldn't be const. Others like `CPipewireStream::GetState` are nice to have const.

File: xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h [hunk 1]
```diff
@@ -65,10 +65,10 @@ class CPipewireNode : public CPipewireProxy
...
   std::unique_ptr<pw_node_info, PipewireNodeInfoDeleter> m_info;
 
-  std::set<spa_audio_format> m_formats;
-  std::set<spa_audio_channel> m_channels;
-  std::set<uint32_t> m_rates;
-  std::set<spa_audio_iec958_codec> m_iec958Codecs;
+  mutable std::set<spa_audio_format> m_formats;
+  mutable std::set<spa_audio_channel> m_channels;
+  mutable std::set<uint32_t> m_rates;
+  mutable std::set<spa_audio_iec958_codec> m_iec958Codecs;
```
Comment (garbear) [inline xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h:71]: Design flaw alarm. This is a lot of state to mark mutable. Usually you only want to use mutable to get around "unrelated" state like mutex internals. The constness of the above methods should agree with these being regular variables.

File: xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h [hunk 2]
```diff
@@ -28,21 +28,21 @@ class CPipewireRegistry;
...
+  pw_node_info* GetInfo() const { return m_info.get(); }
 
-  std::set<spa_audio_format>& GetFormats() { return m_formats; }
-  std::set<spa_audio_channel>& GetChannels() { return m_channels; }
-  std::set<uint32_t>& GetRates() { return m_rates; }
-  std::set<spa_audio_iec958_codec>& GetIEC958Codecs() { return m_iec958Codecs; }
+  std::set<spa_audio_format>& GetFormats() const { return m_formats; }
+  std::set<spa_audio_channel>& GetChannels() const { return m_channels; }
+  std::set<uint32_t>& GetRates() const { return m_rates; }
+  std::set<spa_audio_iec958_codec>& GetIEC958Codecs() const { return m_iec958Codecs; }
```
Comment (garbear) [inline xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h:42]: `const` function should return `const` reference. non-const function can return non-const reference. For example:

```c++
  const std::set<spa_audio_format>& GetFormats() const { return m_formats; }
  std::set<spa_audio_format>& GetFormats() { return m_formats; }
```

File: xbmc/cores/AudioEngine/Sinks/pipewire/PipewireStream.cpp [hunk 3]
```diff
@@ -85,17 +85,17 @@ bool CPipewireStream::TriggerProcess() const
   return true;
 }
 
-void CPipewireStream::Flush(bool drain)
+void CPipewireStream::Flush(bool drain) const
 {
   pw_stream_flush(m_stream.get(), drain);
 }
```
Comment (garbear) [inline xbmc/cores/AudioEngine/Sinks/pipewire/PipewireStream.cpp:91]: A key that this function shouldn't be const is that it immediately calls a C function. C functions can't be marked const, so you shouldn't pretend to the outside world that they are.

Besides, flushing is a mutating operation, so const doesn't really apply here.

**Comment by garbear** [comment]
I left some feedback explaining how I approach consticity.

File: xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h (hunk 2 above)
Comment (garbear) [reply xbmc/cores/AudioEngine/Sinks/pipewire/PipewireNode.h:42]: I find that when functions are documented, it's better to group by function, e.g. https://github.com/xbmc/xbmc/blob/master/xbmc/games/controllers/types/ControllerNode.h

And when functions are undocumented, it's cleaner to group by constness, e.g. https://github.com/garbear/xbmc/blob/retroengine/xbmc/games/controllers/input/ControllerState.h

File: xbmc/cores/AudioEngine/Sinks/pipewire/PipewireStream.cpp (hunk 3 above)
Comment (neo1973) [reply xbmc/cores/AudioEngine/Sinks/pipewire/PipewireStream.cpp:91]: I wouldn't be so strict regarding C functions. E.g. `CPipewireStream::GetState` calls `pw_stream_get_state` ([imp](https://github.com/PipeWire/pipewire/blob/2d5e0ef47ba51a15aef062afb3a2333f8ba90c3d/src/pipewire/stream.c#L1666-L1672)) which is absolutely fine to be used in a const method. (I would have preferred the `struct pw_stream*` parameter to be const but C devs often don't seem to care...)

Comment (garbear) [reply xbmc/cores/AudioEngine/Sinks/pipewire/PipewireStream.cpp:91]: Fair.

> but C devs often don't seem to care

I should have finished with this disclaimer. Dependencies are usually found to be less const-correct, so it's not a reliable key. Indeed, what's important is the context, and getting state would be expected const, and flushing would be  expected non-const.

**Pull request closed without merging**